### ISSUE TITLE
Provide a way to send Liberty logs to OpenTelemetry

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#27711]
## Description

We need a way for users to be able to direct their Liberty logs to OpenTelemetry.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Externally raised requests for enhancements: 
  - [Aha](https://cloud-platform.ideas.aha.io/): Link to the Aha idea (also add a link to this issue in the Aha idea)
    - Feature owner adds label `Aha idea`
  - [Open Liberty Feature Request](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=): Link to the OL GH issue (also add a link to this issue in the Feature Request GH issue)
    - Feature owner adds label `Requested feature`
- UFO: https://ibm.box.com/s/ihoa8t3h18e8w2t41gdcuveo90q24mrd
- FTS: https://github.com/OpenLiberty/open-liberty/issues/29355
- Beta Blog: https://github.com/OpenLiberty/open-liberty/issues/29332
- GA Blog: https://github.com/OpenLiberty/open-liberty/issues/29551

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the feature owner or a delegate of the feature owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The [OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Feature owner adds label `Prioritization - Requested`
  - Feature added to "New" column of the Open Liberty Roadmap. ([OpenLiberty/project-manager](https://github.com/orgs/OpenLiberty/teams/project-manager))
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting
  - `Prioritization - Requested` label removed ([OpenLiberty/project-manager](https://github.com/orgs/OpenLiberty/teams/project-manager) or feature owner)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.  Furthermore, each identified item places a blocking requirement on another team so it must be identified early in the process.  The feature owner may check-off the item if they know it doesn't apply, but otherwise they should work with the focal point to determine what work, if any, will be necessary and make them aware of it.

### **Design Preliminaries**
- [x] UI requirements identified, or N/A. (Feature owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] Accessibility requirements identified, or N/A. (Feature owner and [Accessibility focal point](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
- [x] ID requirements identified, or N/A. (Feature owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability requirements identified, or N/A. (Feature owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT requirements identified, or N/A. (Feature owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified, or N/A. (Feature owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review requested.
  - Feature owner adds label `Design Review Request`
- [x] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] POC Design / UFO approval requested.
  - Feature owner adds label `Design Approval Request`
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [x] No Design requested.
  - Feature owner adds label `No Design Approval Request`
- [x] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`
- [x] Feature / Capability stabilization or discontinuation or N/A
  - Feature owner adds label `Product Management Approval Request` and notifies [OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management)
  - Approver adds label `Product Management Approved` ([OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management))
  - Note: For stabilized, superseded, and discontinued feature/capability, skip the [Beta](#beta) section of the template (you may delete it).  Otherwise, proceed as normal.

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the `Design Approved` or `No Design Approved` label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  All items below **MUST** be completed before beta & GA is requested.

### **Innovation** (Complete 1 week before Beta & GA Feature Complete Date)
- [x] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

### **Legal** (Complete before Beta & GA Feature Complete Date)
- [x] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).

### **Translation** (Complete by Beta & GA Feature Complete Date)
- [x] PII (Program Integrated Information) updates are merged (i.e. all English strings due for translation have been delivered), or N/A.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] [Beta fence](https://github.com/OpenLiberty/open-liberty/wiki/Beta-Fencing) the functionality
  - E.g. `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete by beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [x] Feature implementation and tests completed.
  - [x] All PRs are merged.
  - [x] All related/child issues are closed.
  - [x] All stop ship issues are completed.
- [x] Legal: all necessary approvals granted.
- [x] Innovation: IP identified and any applicable disclosures submitted
- [x] Translation: Feature may only proceed to GA if it has either `Translation - Complete` or `Translation - Missing` label
  - If all translation has been delivered to `release` branch, feature owner adds label `Translation - Complete`.
  - If missing translation does not cause a break in functionality, nor a security or production outage risk, feature owner adds label `Translation - Missing`.
    - Once all missing translations are delivered, the `Translation - Missing` label is replaced with `Translation - Complete`.
  - If missing translation could cause a break in functionality or a security or production outage risk, feature owner adds the `Translation - Blocked` label.
    - Features with `Translation - Blocked` may **NOT** proceed to GA until the label has been replaced with either `Translation - Missing` or `Translation - Complete`.
  - For further guidance, contact [Globalization focal point](https://github.com/orgs/OpenLiberty/teams/globalization-approvers) or the [Release Architect](https://github.com/orgs/OpenLiberty/teams/release-architect).
- [x] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [x] **APIs/Externals** - Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [x] **Demo** - Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI [Iteration Number]` to this issue.
  - Approver adds label `focalApproved:demo`.
- [x] **FAT** - All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.

### **Design Approved Features**
- [x] **ID** - Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **InstantOn** - InstantOn capable or N/A. ([OpenLiberty/instantOn-approvers](https://github.com/orgs/OpenLiberty/teams/instantOn-approvers))
  - Approver adds label `focalApproved:instantOn`.
- [x] **Performance** - Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [x] **Serviceability** - Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [x] **STE** - Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [x] **SVT** - System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [ ] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Friday after GM)
- [ ] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GM** (Complete before GA)
- [ ] After confirming this feature has been included in the GM driver, feature owner closes this issue.

### **Post GA**
- [ ] Remove the `target:ga` and `target:YY00X` labels, and add the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **Standalone Feature Blog Post** - A blog post specifically about your feature or N/A.  (Feature owner and [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** - OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** - Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by pgunapal** [comment]
### Design [DRAFT] : 
*** Subject to change ***

**High Level User Story:**
As an Operations engineer, I want to be able to export logs from Open Liberty to an Open Telemetry Exporter.

OpenTelemetry defines a Logs Bridge API for emitting LogRecords. OpenTelemetry provides a [Logs Bridge API](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/) and [SDK](https://opentelemetry.io/docs/specs/otel/logs/sdk/), which can be used together with existing logging libraries to automatically inject the trace context in the emitted logs, and provide an easy way to send the logs via OTLP. Instead of modifying each logging statement, log appenders use the API to bridge logs from existing logging libraries to the OpenTelemetry data model, where the SDK controls how the logs are processed and exported. The typical log SDK configuration installs a log record processor and exporter.

 The LogRecordProcessor from the [Logs SDK](https://opentelemetry.io/docs/languages/java/instrumentation/#logs-sdk) allows us to process and decorate the LogRecord fields to map to OTel Log Data Model.

- **SimpleLogRecordProcessor:** This is an implementation of LogRecordProcessor which passes finished logs and passes the export-friendly ReadableLogRecord representation to the configured LogRecordExporter, as soon as they are finished.

- **BatchLogRecordProcessor:** This is an implementation of the LogRecordProcessor which create batches of LogRecords and passes the export-friendly ReadableLogRecord representations to the configured LogRecordExporter.

BatchLogRecordProcessor and SimpleLogRecordProcessor are paired with [LogRecordExporter](https://opentelemetry.io/docs/languages/java/instrumentation/#logrecord-exporter), which is responsible for sending telemetry data to a particular backend.

### Feature Design:
In Open Liberty, Open Telemetry is initialized using the SDK autoconfiguration extension,  instead of manually creating the OpenTelemetry instance by using the SDK builders directly in the code. This approach allows you to autoconfigure the OpenTelemetry SDK based on a standard set of supported environment variables and system properties. Hence, the logging providers can be configured using environment variables.
Ref: https://opentelemetry.io/docs/languages/java/instrumentation/#autoconfiguration

Since, the mpTelemetry-2.0 OpenTelemetry instance is dependent on application thread context, it will be difficult to get the instance, when we are not apart of the application thread context, such as during server start up. Hence, we need to explicitly create a new server-level Open Telemetry instance, which would have its own server-specific OTel configuration. This will also work for the multi-app scenarios as well.

- Enable the `mpTelemetry-2.0` feature and configure the Environment variables to enable the server-level OTel SDK and configure the OTel Log exporter and LogRecord Processors, as follows:
```
mp.server.otel.service.name (can be similar as app-level config)
mp.server.otel.sdk.disabled=false
mp.server.otel.logs.exporter=otlp
mp.server.otel.exporter.otlp.endpoint=http://localhost:4317/
```
There will be server-level resource attributes configured for the server-level OpenTelemetry instance, as well.
(https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#exporter-selection)

By default, the SimpleLogRecordProcessor will be enabled, where the records will be send immediately. However, if you want to send the records in batches, you can also configure the following logging specific Batch LogRecord Processor Environment variables to configure how often and how to export the logs over, and as well as log record limits for attributes.
(https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#batch-logrecord-processor)
```
OTEL_BLRP_SCHEDULE_DELAY : Delay interval (in milliseconds) between two consecutive exports (Default = 1000)
OTEL_BLRP_EXPORT_TIMEOUT : Maximum allowed time (in milliseconds) to export data (Default = 30000)
OTEL_BLRP_MAX_QUEUE_SIZE : Maximum queue size (Default = 2048)
OTEL_BLRP_MAX_EXPORT_BATCH_SIZE : Maximum batch size (Default = 512)

OTEL_LOGRECORD_ATTRIBUTE_VALUE_LENGTH_LIMIT : Maximum allowed attribute value size (Default = no limit)
OTEL_LOGRECORD_ATTRIBUTE_COUNT_LIMIT : Maximum allowed log record attribute count (Default = 128)
```

- Use the mpTelemetry-2.0 feature in server.xml, with server configuration attribute to configure the different log sources (message, trace, accessLog, ffdc, garbageCollection, and audit) we want to send to OTel exporter. Default value for the `logSources` attribute would be `message`.

```
<feature>mpTelemetry-2.0</feature>
…
<mpTelemetry logSources=“message, trace, accessLog, ffdc, audit”/>
```
- The feature design would be similar to the Logstash Collector 1.0 feature, where we will be subscribing a “OTel Handler” to the collector manager, where we would get access to each event and format the event accordingly (JSONify), to map to the OTel Log Data Model and then emit it to the “Target”, which would be the configured exporter for OTel logs.

- The corresponding SpanID and TraceID for trace context and application will be retrieved from the LogRecordContext (ext_traceID and ext_spanID).

**Mapping Open Liberty Log Record to Open Telemetry Logs Data Model**
(https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition)
Note: When formatting the event, JSONify the event, so the event is structured properly.
```
Open Liberty Log Record | Open Telemetry Logs Data Model
=============================================
ibm_datetime = Timestamp
ext_traceId = TraceId
ext_spanId = SpanId
loglevel = SeverityText
(Refer to table below) = SeverityNumber 
message  = Body *** Should be the entire JSON payload instead?
host = Resource[“host.name”]
service.name = Resource[“service.name”]
io.openliberty.microprofile.telemetry = InstrumentationScope
Map the rest of the fields as Key:Value pairs = Attributes[“Key”]
Throwable example snippet:
throwable.getClass().getName() = Attributes[SemanticAttributes.EXCEPTION_TYPE]
throwable.getMessage = Attributes[SemanticAttributes.EXCEPTION_MESSAGE]
throwable.printStackTrace() = Attributes[SemanticAttributes.EXCEPTION_STACKTRACE]
```

**Log Level Mapping to Open Telemetry Severity Text**
(https://opentelemetry.io/docs/specs/otel/logs/data-model/#severity-fields)
(https://opentelemetry.io/docs/specs/otel/logs/data-model-appendix/#appendix-b-severitynumber-example-mappings)

```
Open Liberty Log Level | Open Telemetry Logs Severity Text / Number
====================================================
FATAL = FATAL / 21
SEVERE = ERROR / 17
WARNING = WARN / 13
AUDIT = INFO2 / 10
INFO = INFO / 9
CONFIG = DEBUG4 / 8
DETAIL = DEBUG3 / 7
FINE = DEBUG2 / 6
FINER = DEBUG / 5
FINEST = TRACE / 1
```

**Comment by pgunapal** [comment]
### High-level Implementation Design Details [DRAFT] : 
*** Subject to change ***

- Update mpTelemetry-2.0 feature files (public and private, for different EE versions) to include project `io.openliberty.microprofile.telemetry.2.0.logging.internal`
- Add the required OpenTelemetry Logs SDK and API libraries to the [mpTelemetry-2.0 feature file](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.websphere.appserver.features/visibility/public/mpTelemetry-2.0/io.openliberty.mpTelemetry-2.0.feature):
```
io.opentelemetry.sdk.logs.export;type="third-party",\
io.opentelemetry.sdk.logs.data;type="third-party",\
io.opentelemetry.api.logs;type="third-party",\
```
- Create a new class `io.openliberty.microprofile.telemetry20.OpenTelemetryHandler`, in the new project `io.openliberty.microprofile.telemetry.2.0.logging.internal`, that extends `com.ibm.ws.collector.Collector`, similar to how its done in [LogstashCollector](https://github.com/OpenLiberty/open-liberty/blob/351c9ab62b21403de7ce0f8a8bd6c68d8160b8a3/dev/com.ibm.ws.logstash.collector/src/com/ibm/ws/logstash/collector/internal/LogstashCollector.java#L55).
- The contents of `OpenTelemetryHandler` class and the BND files should be similar to the [LogstashCollector project](https://github.com/OpenLiberty/open-liberty/tree/integration/dev/com.ibm.ws.logstash.collector), however, omit the redundant components, such as SSL, ExecutorService, which will not be needed for this feature.
- Add the corresponding mpTelemetry-2.0 packages to the BND file.
```
Import-Package:
io.openliberty.microprofile.telemetry.internal.common

-buildpath:
	io.openliberty.microprofile.telemetry.internal.common;version=latest,\
	io.openliberty.io.opentelemetry.2.0;version=latest
```
- Ensure the correct metatype is defined for the server configuration of mpTelemetry-2.0 (e.g. `logSources`)
- Update the [OpenTelemetryVersionedConfigurationImpl](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.microprofile.telemetry.2.0.internal/src/io/openliberty/microprofile/telemetry20/internal/config/OpenTelemetryVersionedConfigurationImpl.java) class file to remove the following lines, since we should be enabling Logs, as part of this feature.
```
telemetryProperties.put(OpenTelemetryConstants.CONFIG_LOGS_EXPORTER_PROPERTY, "none");
telemetryProperties.put(OpenTelemetryConstants.ENV_LOGS_EXPORTER_PROPERTY, "none");
```
- In the `OpenTelemetryHandler.activate()` method, retrieve and set the server-level OpenTelemetryInfo object. It will be using the [OpenTelemetryAccessor](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/interfaces/OpenTelemetryAccessor.java) interface from the `io.openliberty.microprofile.telemetry.internal.common` project (TBD - details to follow from MP Telemetry team)
- Get the configured OpenTelemetry LogProvider by calling the `OpenTelemetryInfo.getOpenTelemetry().getLogsBridge()`. (https://javadoc.io/doc/io.opentelemetry/opentelemetry-api/latest/io/opentelemetry/api/OpenTelemetry.html)
- In the `OpenTelemetryHandler.formatEvents()` method, by using the previously retrieved LogProvider, get the LogBuilder (https://javadoc.io/doc/io.opentelemetry/opentelemetry-api/latest/io/opentelemetry/api/logs/LoggerProvider.html), with the configured instrumentation name, and use the builder to map the Open Liberty event records to the appropriate OpenTelemetry Log Data model. 
- Once, the builder is mapped with the corresponding fields, call `builder.emit()` to export the logs to the exporter.

Below is a generic high-level code snippet on how to retrieve the OpenTelemetry LogProvider/builder, and to map generic JUL Log Record fields to Open Telemetry Log Data Model, and then to export it to the configured exporter.

Snippet is from the [Open Telemetry JUL Java agent instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/java-util-logging/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jul/JavaUtilLoggingHelper.java) :

```
...
String instrumentationName = logger.getName();
    if (instrumentationName == null || instrumentationName.isEmpty()) {
      instrumentationName = "ROOT";
    }
    LogRecordBuilder builder =
        GlobalOpenTelemetry.get()
            .getLogsBridge()
            .loggerBuilder(instrumentationName)
            .build()
            .logRecordBuilder();
    mapLogRecord(builder, logRecord);
    builder.emit();

private static void mapLogRecord(LogRecordBuilder builder, LogRecord logRecord) {
    // message
    String message = FORMATTER.formatMessage(logRecord);
    if (message != null) {
      builder.setBody(message);
    }

    // time
    long timestamp = logRecord.getMillis();
    builder.setTimestamp(timestamp, TimeUnit.MILLISECONDS);

    // level
    Level level = logRecord.getLevel();
    if (level != null) {
      builder.setSeverity(levelToSeverity(level));
      builder.setSeverityText(logRecord.getLevel().getName());
    }

    AttributesBuilder attributes = Attributes.builder();

    // throwable
    Throwable throwable = logRecord.getThrown();
    if (throwable != null) {
      attributes.put(SemanticAttributes.EXCEPTION_TYPE, throwable.getClass().getName());
      attributes.put(SemanticAttributes.EXCEPTION_MESSAGE, throwable.getMessage());
      StringWriter writer = new StringWriter();
      throwable.printStackTrace(new PrintWriter(writer));
      attributes.put(SemanticAttributes.EXCEPTION_STACKTRACE, writer.toString());
    }

    if (captureExperimentalAttributes) {
      Thread currentThread = Thread.currentThread();
      attributes.put(SemanticAttributes.THREAD_NAME, currentThread.getName());
      attributes.put(SemanticAttributes.THREAD_ID, currentThread.getId());
    }

    builder.setAllAttributes(attributes.build());

    // span context
    builder.setContext(Context.current());
  }
...
```

Note: OpenTelemetry have implemented Log Appenders [Log4J](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/src/main/java/io/opentelemetry/instrumentation/log4j/appender/v2_17/internal/LogEventMapper.java#L104) and [Logback](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/logback/logback-appender-1.0/library/src/main/java/io/opentelemetry/instrumentation/logback/appender/v1_0/internal/LoggingEventMapper.java).

**Comment by benjamin-confino** [comment]
Hello. 

> In the OpenTelemetryHandler.activate() method, retrieve and set the server-level OpenTelemetryInfo object. It will be using the [OpenTelemetryAccessor](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/interfaces/OpenTelemetryAccessor.java) interface from the io.openliberty.microprofile.telemetry.internal.common project (TBD - details to follow from MP Telemetry team)

This should work fine but there is one hidden gotcha to be aware off. In OpenTelemetryInfoFactory we have a check when an OpenTelemetryInfo is created `if (j2EEName.startsWith("io.openliberty") || j2EEName.startsWith("com.ibm.ws")) {` you will need to make sure this if statement returns false when OpenTelemetryHandler calls us. I suspect without modification it will be true.

**Comment by pgunapal** [comment]
Thanks @benjamin-confino ! Good point, will make sure that doesn't break for us. Right, `getOpenTelemetryInfo()` will be called by internal code, so it would be true.

**Comment by Yas-Franco** [comment]
POC notes from July 15th: 
- [x] Slide 9: Update the slide to mention the existing behaviours in Open Liberty today 
- [x] Slide 10: Do we lose some context when the agent is on?
- [x] Slide 11: Are we losing with ibm_ and ext_ logs when the agent is on? It would be helpful to see another column with what the agent produces compared to what we produce. 
- [x] Slide 13: Where did you get the severity numbers from? Will we have liberty documentation the logging level mapping? 
- [x] Slide 14: Do we still prefix ibm_ in the attributes in all of the logs? 
- [x] Slide 14: Should the attributes added by ibm start with ibm_ to make it clear that they are not otel semantic conventions?
- [x] Slide 20: How do you turn off logs? otel.logs.exporter would need to be none and shown in this slide. 
- [x] Slide 20: What is the default for <mpTelemery source= />
- [x] Slide 21: Are the attributes/ values case sensitive? 
- [x] Slide 21: When you set the exporter to console, what happens to the console logs? 
- [x] Slide 21: The configuration cannot be updated dynamically. What is important for deployment time and containerisation time? 
- [x] Slide 26: Instead of clearing earlier messages, they would not be added to the queue. 
- [x] Slide 29: The configuration is not part of the beta. 
- [x] Slide 30: Negative tests should also be added. 
- [x] Slide 31: We should do a 1-3 day run for these tests. 
- [x] Slide 32: Is the data loss correct for batch and asynchronous processors? 
- [x] Slide 32: Setting the otel.logs.export to none would still create an otel instance. 
- [x] Slide 33: Batch is used as default but has a negative effect on performance. This should be looked in to. 
- [x] Slide 34: Would the certificate be built before the restore point? 

Further point at end of UFO: 
Ensure that negative cases are well tested and that instantOn is considered when testing the feature.

**Comment by pgunapal** [comment]
@yasmin-aumeeruddy Thank you for the notes, I have updated the UFO with the comments from the socialization.

**Comment by NottyCode** [comment]
Slide 9 - I think we need an epic for access logs and audit. I understand why mapping these log events is non-trivial and less important, but we should do it, especially for access logs.
Slide 12 - It is odd that we don't map anything into the body for FFDC. This is raising my "something wrong" thing.

**Comment by pgunapal** [comment]
@NottyCode 
For Slide 9, we have opened two epics to address access and audit logs
- https://github.com/OpenLiberty/open-liberty/issues/29228
- https://github.com/OpenLiberty/open-liberty/issues/29229

For Slide 12, we decided to map the exception message from the triggered event to the body, in addition to Semantic Convention Attribute name (`exception.message`).

I have updated the UFO with the above.

**Comment by pgunapal** [comment]
@OpenLiberty/demo-approvers Demo scheduled for EOI 24.17

**Comment by pgunapal** [comment]
@OpenLiberty/id-approvers ID Doc Issue opened: https://github.com/OpenLiberty/docs/issues/7459

**Comment by pgunapal** [comment]
@OpenLiberty/externals-approvers Can you please review the approval for this feature, there are no exposed public APIs as part of this particular feature.

**Comment by chirp1** [comment]
Slack with Prashanth, David, Ram, me. Prashanth provided the necessary info in the following doc issue: https://github.com/OpenLiberty/docs/issues/7459. I approved the feature.

**Comment by tngiang73** [comment]
@pgunapal: WASWIN is good with the STE slides. STE approved.

**Comment by donbourne** [comment]
OL:

**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. IBM Support, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that IBM Support should be able to quickly address those problems without need to engage SMEs?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that IBM Support should be able to quickly address those problems without need to engage SMEs?

4.  Which IBM Support / SME queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective IBM Support/SME teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by donbourne** [comment]
@clarkek123 will be handling the serviceability approval for this epic.

**Comment by dave-waddling** [comment]
Thanks for completing the FTS. The results of the mini-SOE look good so adding the FAT Focal approval.

**Comment by pgunapal** [comment]
@clarkek123 I have filled out the template below, can you please review the Serviceability approval for this feature?

Serviceability Approval Comment - Please answer the following questions for serviceability approval:

UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

A: Yes. For the logging component of the `mpTelemetry-2.0`, the logs should be exported automatically if the `mpTelemetry-2.0` feature is enabled in the server.xml, along with the `otel.sdk.disabled=true` is configured either in Liberty or the application, and the `otel.logs.exporter` is set to a valid exporter.  Below are some of the common user error scenarios for logs not being exported that we tested in FATs, as well as manually. 
- `otel.logs.exporter` is set to none. – A warning message is shown stating that the exporter for logs is disabled.
- `otel.sdk.disabled` is not set to false. –  A warning message is shown that the OpenTelemetry SDK instance is not enabled.
- The `source` attribute is empty in the server.xml. (<mpTelemetry source=“”/>), a warning message will be shown.
- If there are invalid log event sources configured, a warning message will be logged:  CWMOT5005W: The MicroProfile Telemetry logging feature does not recognize the [ {0} ] log source.  

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. IBM Support, test team, or another development team).
a) What problem paths were tested and demonstrated?
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that IBM Support should be able to quickly address those problems without need to engage SMEs?

A:
a) The problems paths mentioned above were also demoed to analyze the problem paths.
b) Local, EOI, SVT, Performance teams.
c) Yes

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that IBM Support should be able to quickly address those problems without need to engage SMEs?

A:
a) Daniel Guinan
b) Yes

Which IBM Support / SME queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective IBM Support/SME teams know they are supporting it. Ask Don Bourne if you need links or more info.

A:
WAS L3: Logging

Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

A: 
No

**Comment by clarkek123** [comment]
Based on the information provided  above for Serviceability showing common error paths testing with demo with approval from Local, SVT and Perfomance teams and SVT signoff on the paths included for serviceability, I have added the Serviceability approval for this feature.

**Comment by pgunapal** [comment]
Closing, as this feature is complete, and delivered for 24.0.0.9.
